// schm/dist/utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.parseValidatorOption = exports.isObject = exports.isFunction = exports.isSchema = exports.isArray = exports.toArray = undefined;

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

var _omit = require("lodash/omit");

var _omit2 = _interopRequireDefault(_omit);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var toArray = exports.toArray = function toArray(value) {
  return [].concat(typeof value === "undefined" ? [] : value);
};

var isArray = exports.isArray = function isArray(value) {
  return Array.isArray(value);
};

var isSchema = exports.isSchema = function isSchema(schema) {
  return !!(schema && schema.params && schema.parsers && schema.validators && schema.parse && schema.validate && schema.merge);
};

var isFunction = exports.isFunction = function isFunction(value) {
  return typeof value === "function";
};

var isObject = exports.isObject = function isObject(value) {
  return (typeof value === "undefined" ? "undefined" : _typeof(value)) === "object";
};

var parseValidatorOption = exports.parseValidatorOption = function parseValidatorOption(option, ignoreArray) {
  var isArrayAndHasMessage = function isArrayAndHasMessage(v) {
    return isArray(v) && v.length === 2 && typeof v[1] === "string";
  };
  var isValidatorObject = function isValidatorObject(v) {
    return v && (v.message || v.msg);
  };

  if (!ignoreArray && isArrayAndHasMessage(option)) {
    return {
      optionValue: option[0],
      message: option[1]
    };
  } else if (isValidatorObject(option)) {
    var optionWithoutMessage = (0, _omit2.default)(option, "message", "msg");
    var optionValueKey = Object.keys(optionWithoutMessage)[0];
    return {
      optionValue: optionWithoutMessage[optionValueKey],
      message: option.message || option.msg
    };
  }
  return { optionValue: option };
};
